### PR TITLE
[target allocator] Restart target allocator when config changes

### LINKED ISSUE #1882: [target allocator] Restart Target Allocator when its configuration changes
We ensure the collector is restarted on configuration change by including a hash of the configuration in an annotation. We don't do this for Target Allocator though - it needs to be manually restarted.

### PR DESCRIPTION
**Pull request by swiatekm-sumo**
Fixes #1882.

As part of these changes, I moved ConfigMap generation for the target allocator to the `targetallocator` package. This seems like the right place for it conceptually, and it made it easier to simply calculate the hash of the ConfigMap content without worrying about which changes in the CR are meaningful. This move is in a separate commit, and it should be easier to review that way.

### CODE REVIEW

**Comment by swiatekm-sumo** [comment]
I believe the E2E failure is transient, but can't restart them myself. :(

File: pkg/targetallocator/configmap.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+	"github.com/open-telemetry/opentelemetry-operator/apis/v1alpha1"
+	"github.com/open-telemetry/opentelemetry-operator/pkg/naming"
+	"github.com/open-telemetry/opentelemetry-operator/pkg/targetallocator/adapters"
+)
+
+const (
+	targetAllocatorFilename = "targetallocator.yaml"
+)
+
+func ConfigMap(instance v1alpha1.OpenTelemetryCollector) (corev1.ConfigMap, error) {
```
Comment (jaronoff97) [inline pkg/targetallocator/configmap.go:34]: Was there anything that changed in here other than just moving the function?

**Review by jaronoff97 — COMMENTED** [review]
Just a few questions I had while reviewing, I think this looks about good to go though.

File: pkg/targetallocator/annotations.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+func Annotations(instance v1alpha1.OpenTelemetryCollector) map[string]string {
+	// Make a copy of PodAnnotations to be safe
+	annotations := make(map[string]string, len(instance.Spec.PodAnnotations))
+	for key, value := range instance.Spec.PodAnnotations {
+		annotations[key] = value
+	}
+
+	configMap, err := ConfigMap(instance)
+	if err == nil {
+		annotations[configMapHashAnnotationKey] = getConfigMapSHA(configMap)
```
[the remark this answers is not in the document]
Comment (swiatekm-sumo) [reply pkg/targetallocator/annotations.go:38]: The E2E test I added proves that we actually do create a new ReplicaSet on config change, resulting in a new rollout.

File: pkg/targetallocator/configmap.go (hunk 1 above)
Comment (swiatekm-sumo) [reply pkg/targetallocator/configmap.go:34]: No, straight up copy + paste for both the function body and the tests.

File: pkg/targetallocator/configmap.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+	// Collector supports environment variable substitution, but the TA does not.
+	// TA ConfigMap should have a single "$", as it does not support env var substitution
+	prometheusReceiverConfig, err := adapters.UnescapeDollarSignsInPromConfig(instance.Spec.Config)
+	if err != nil {
+		return corev1.ConfigMap{}, err
+	}
+
+	taConfig := make(map[interface{}]interface{})
+	taConfig["label_selector"] = map[string]string{
+		"app.kubernetes.io/instance":   fmt.Sprintf("%s.%s", instance.Namespace, instance.Name),
```
[the remark this answers is not in the document]
Comment (swiatekm-sumo) [reply pkg/targetallocator/configmap.go:53]: We actually do as it turns out, I'll make use of it here.

[the remark this answers is not in the document]
Comment (swiatekm-sumo) [reply pkg/targetallocator/configmap.go]: We do have it, but using it causes a change to the labels, see: https://github.com/open-telemetry/opentelemetry-operator/pull/1889/commits/427ffed65fc2c8f9269056f900f41cf80e58b34e.

File: pkg/targetallocator/annotations.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+// Annotations returns the annotations for the TargetAllocator Pod.
+func Annotations(instance v1alpha1.OpenTelemetryCollector) map[string]string {
+	// Make a copy of PodAnnotations to be safe
+	annotations := make(map[string]string, len(instance.Spec.PodAnnotations))
+	for key, value := range instance.Spec.PodAnnotations {
+		annotations[key] = value
+	}
+
+	configMap, err := ConfigMap(instance)
```
[the remark this answers is not in the document]
Comment (swiatekm-sumo) [reply pkg/targetallocator/annotations.go:36]: At what level would you want to calculate it. I believe Deployments and ConfigMaps are reconciled reparately, so there's no way to only do it once. The current solution is dumb, but it more or less guarantees correctness, and it's nicely self-contained.

File: pkg/targetallocator/configmap.go (hunk 1 above)
Comment (jaronoff97) [reply pkg/targetallocator/configmap.go:34]: cool, just checking :)

File: pkg/targetallocator/annotations.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (jaronoff97) [reply pkg/targetallocator/annotations.go:36]: okay, that's fine just wanted to call it out and see if there's a better option.

**Review by jaronoff97 — APPROVED** [review]

**Comment by jaronoff97** [comment]
any other @open-telemetry/operator-ta-maintainers want to take a look?

**Review by secustor — APPROVED** [review]

**Comment by swiatekm-sumo** [comment]
Do I need to do anything specific for Prow CI to pick up this PR?

**Pull request merged**